src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.Utf8.cs

```diff
@@ -0,0 +1,332 @@
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+using System.Text;
+using System.Text.Unicode;
+
+namespace System.Globalization
+{
+    internal static partial class Ordinal
+    {
+        internal static int CompareStringIgnoreCaseUtf8(ref byte strA, int lengthA, ref byte strB, int lengthB)
+        {
+            int length = Math.Min(lengthA, lengthB);
+            int range = length;
+
+            ref byte charA = ref strA;
+            ref byte charB = ref strB;
+
+            const byte maxChar = 0x7F;
+
+            while ((length != 0) && (charA <= maxChar) && (charB <= maxChar))
+            {
+                // Ordinal equals or lowercase equals if the result ends up in the a-z range
+                if (charA == charB ||
+                    ((charA | 0x20) == (charB | 0x20) && char.IsAsciiLetter((char)charA)))
+                {
+                    length--;
+                    charA = ref Unsafe.Add(ref charA, 1);
+                    charB = ref Unsafe.Add(ref charB, 1);
+                }
+                else
+                {
+                    int currentA = charA;
+                    int currentB = charB;
+
+                    // Uppercase both chars if needed
+                    if (char.IsAsciiLetterLower((char)charA))
+                    {
+                        currentA -= 0x20;
+                    }
+                    if (char.IsAsciiLetterLower((char)charB))
+                    {
+                        currentB -= 0x20;
+                    }
+
+                    // Return the (case-insensitive) difference between them.
+                    return currentA - currentB;
+                }
+            }
+
+            if (length == 0)
+            {
+                return lengthA - lengthB;
+            }
+
+            range -= length;
+
+            return CompareStringIgnoreCaseNonAsciiUtf8(ref charA, lengthA - range, ref charB, lengthB - range);
+        }
+
+        internal static int CompareStringIgnoreCaseNonAsciiUtf8(ref byte strA, int lengthA, ref byte strB, int lengthB)
+        {
+            // NLS/ICU doesn't provide native UTF-8 support so we need to convert to UTF-16 and compare that way
+
+            // Convert strA using stackalloc for <= 256 characters and ArrayPool otherwise
+
+            char[]? strAUtf16Array;
+            scoped Span<char> strAUtf16;
+            int strAMaxCharCount = Encoding.UTF8.GetMaxCharCount(lengthA);
+
+            if (strAMaxCharCount <= 256)
+            {
+                strAUtf16Array = null;
+                strAUtf16 = stackalloc char[512];
+            }
+            else
+            {
+                strAUtf16Array = ArrayPool<char>.Shared.Rent(strAMaxCharCount);
+                strAUtf16 = strAUtf16Array.AsSpan(0, strAMaxCharCount);
+            }
+
+            int strAUtf16Length = Encoding.UTF8.GetChars(MemoryMarshal.CreateReadOnlySpan(ref strA, lengthA), strAUtf16);
```

Comment:
`Encoding.UTF8.GetChars` is a lossy operation and could give you incorrect output. For example, the input byte arrays `new byte[] { 0xFE }` and `new byte[] { 0xFF }` _should not_ compare as equal under any UTF-8 comparer, but `Encoding.UTF8.GetChars` will turn both inputs into the output `new char[] { 0xFFFD }`, which _will_ compare as equal.

The logic you want is instead something along these lines:

* Call `Rune.DecodeFromUtf8` on both inputs.
    * If _both_ return values are `OperationStatus.Completed`, then convert each to uppercase and compare the uppercase values against one another. Fail if the uppercase values do not match.
    * Else, if _both_ return values are `OperationStatus.<AnythingElse>`, then look at the _bytesConsumed_ out value of each `DecodeFromUtf8` call and look at the next (that number of bytes) from each input, comparing them for bytewise equality. Fail if the bytewise equality check does not pass.
    * Otherwise, if the `OperationStatus` values don't match, fail immediately.
* Slice each input by the number of bytes processed and loop.

Not only does this guarantee correct output - it guarantees that we're bounding the operation to `O(i)`, where _i_ is the index of the first non-matching character. This means that if the caller is invoking this method in a loop where each iteration of the loop slices, the caller is guaranteed `O(n)` aggregate performance in the worst-case. If we perform an entire transcoding operation before doing the equality check, then the caller might end up in an `O(n^2)` loop, which could lead to performance or security issues when the parser deals with untrusted input.

Reply:
Updated to follow this logic.